tools/create-hybrid-image

```diff
@@ -0,0 +1,312 @@
+      -d "Debian 10 (bionic) 64-bit image with just essential packages... " \\
+      -u https://docs.joyent.com/images/hvm/debian
+
+OPTIONS:
+  -i The ISO
+  -n The name of the image as it would appear in the manifest
+  -d The description of the image
+  -o Directory in which to place output.  Default is ".".
+  -p Prototype manifest file
+  -u The homepage URL for the image
+  -h Show this message
+
+EOF
+}
+
+while getopts "hi:n:d:o:p:u:" OPTION; do
+	case $OPTION in
+	h)
+		usage
+		exit 0
+		;;
+	i)
+		ISO=${OPTARG}
+		;;
+	n)
+		IMAGE_NAME=${OPTARG}
+		;;
+	d)
+		DESC=${OPTARG}
+		;;
+	o)
+		OUTPUT_DIR=${OPTARG}
+		;;
+	p)
+		PROTOTYPE=${OPTARG}
+		;;
+	u)
+		HOMEPAGE=${OPTARG}
+		;;
+	?)
+		usage 1>&2
+		exit 1
+		;;
+	esac
+done
+
+if [[ -z ${ISO} || -z ${IMAGE_NAME} || -z ${DESC} || -z ${HOMEPAGE} || \
+    -z ${PROTOTYPE} ]]; then
+	echo "FATAL: All of -i, -n, -d, -p, and -u are required." 1>&2
+	usage 1>&2
+	exit 1
+fi
+
+if [[ ! -f $PROTOTYPE ]]; then
+	echo "FATAL: Manifest prototype file '$PROTOTYPE' does not exist" 1>&2
+	exit 1
+fi
+
+shift $((OPTIND - 1))
+if (( $# == 0 )); then
+	# Set to a scalar value that will work for a test later.  bash is
+	# unhappy if QEMU_ARGS is declared as an empty array and it is used.
+	QEMU_ARGS=
+else
+	QEMU_ARGS=( "$@" )
+fi
+
+PIDFILE=/tmp/qemu-$IMAGE_NAME.pid
+
+function sanity_check
+{
+	sane=true
+
+	topds=$(zfs list -Ho name /data || true)
+	if [[ -z "$topds" ]]; then
+		sane=false
+		topds=zones/$(zonename)/data
+		echo "$0: No zfs filesystem mounted at /data" 1>&2
+		echo "" 1>&2
+		echo "Fix by running the following in the global zone:" 1>&2
+		echo "  zfs create -o zoned=on -o mountpoint=/data $topds" 1>&2
+		echo "  zonecfg -z $(zonename) 'add dataset; set name=$topds;" \
+		    "end'" 1>&2
+		echo "" 1>&2
+	fi
+
+	if [[ ! -f /smartdc/bin/qemu-system-x86_64 ]]; then
+		sane=false
+		echo "$0: qemu executable not present." 1>&2
+		echo "" 1>&2
+		echo "Fix by running the following in the global zone:" 1>&2
+		echo "  zonecfg -z $(zonename) 'add fs; set type=lofs;" \
+		    "set dir=/smartdc; set special=/smartdc;" \
+		    "set options=ro; end'" 1>&2
+		echo "" 1>&2
+	fi
+
+	if [[ ! -c /dev/kvm ]]; then
+		sane=false
+		echo "$0: kvm device not present." 1>&2
+		echo "" 1>&2
+		echo "Fix by running the following in the global zone:" 1>&2
+		echo "  zonecfg -z $(zonename) 'add device; set match=kvm;" \
+		    "end'" 1>&2
+		echo "" 1>&2
+	fi
+
+	if [[ $sane == false ]]; then
+		echo "$0: Reboot the zone after making the changes described" \
+		    "above." 1>&2
+		exit 1
+	fi
+}
+
+function create_blank {
+	echo -n "==> Creating blank $IMAGE_NAME virtual disk..."
+
+	# /dev/zvol is buggy in zones, sometimes leaving ghost entries around
+	# hiding newer ones of the same name.  Using a random UUID in each zvol
+	# name avoids this problem.
+	DISKVOL=$topds/$IMAGE_NAME.$VM_UUID.disk0
+	if zfs list "$DISKVOL" >/dev/null 2>&1; then
+		pfexec zfs destroy -r "$DISKVOL"
```

Comment:
Should we check for the Primary Adminstrator profile early on in the script instead of only blowing up at this point?